cli/src/main.rs

```diff
@@ -110,3 +145,40 @@ fn dot_federated_graph(file_paths: &[PathBuf]) -> Result<(), FederationError> {
     println!("{}", query_graph::output::to_dot(&query_graph));
     Ok(())
 }
+
+fn cmd_validate(file_paths: &[PathBuf]) -> Result<(), FederationError> {
+    load_supergraph(file_paths)?;
+    println!("[SUCCESS]");
+    Ok(())
+}
+
+fn cmd_compose(file_paths: &[PathBuf]) -> Result<(), FederationError> {
+    let supergraph = compose_files(file_paths)?;
+    println!("{}", supergraph.schema.schema());
+    Ok(())
+}
+
+fn cmd_extract(file_path: &PathBuf, dest: Option<PathBuf>) -> Result<(), FederationError> {
+    let supergraph = load_supergraph_file(file_path)?;
+    let subgraphs = supergraph.extract_subgraphs()?;
+    if let Some(ref dest) = dest {
+        fs::create_dir_all(dest).map_err(|_| SingleFederationError::Internal {
+            message: "Error: directory creation failed".into(),
+        })?;
+        for (name, subgraph) in subgraphs {
+            let subgraph_path = dest.join(format!("{}.graphql", name));
+            fs::write(subgraph_path, subgraph.schema.schema().to_string()).map_err(|_| {
+                SingleFederationError::Internal {
+                    message: "Error: file output failed".into(),
+                }
+            })?;
```

Comment:
We have a constructor function for internal federation errors:
```suggestion
            fs::write(subgraph_path, subgraph.schema.schema().to_string()).map_err(|_| {
                FederationError::internal("Error: file output failed")
            })?;
```

Reply:
Currently, `FederationError::internal` is `pub(crate)` and not visible to CLI. Shall we open it up to public?